lib/graphql/subscriptions/action_cable_subscriptions.rb

```diff
@@ -82,23 +82,22 @@ class Subscriptions
     #   end
     #
     class ActionCableSubscriptions < GraphQL::Subscriptions
-      SUBSCRIPTION_PREFIX = "graphql-subscription:"
-      EVENT_PREFIX = "graphql-event:"
-
       # @param serializer [<#dump(obj), #load(string)] Used for serializing messages before handing them to `.broadcast(msg)`
-      def initialize(serializer: Serialize, **rest)
+      def initialize(serializer: Serialize, subscription_prefix: 'graphql-subscription:', event_prefix: 'graphql-event:', **rest)
```

Comment:
Why accept two options here? Is there ever a time that an application would need to specify both? 

What if, instead, we accepted: 

```
prefix: ""
```

And added that to `SUBSCRIPTION_PREFIX` and `EVENT_PREFIX` under the hood? 

That way, each schema could specify its prefix, and they'd be in different namespaces, eg: 

```ruby 
use GraphQL::Subscriptions::ActionCableSubscriptions, prefix: "foo"
# ... 
use GraphQL::Subscriptions::ActionCableSubscriptions, prefix: "bar"
```

Would that do the job?

Reply:
> Why accept two options here? Is there ever a time that an application would need to specify both?

There were 2 independent constants so I moved them "brainlessly" into 2 independent kwargs without thinking much to keep it natural to anyone who would know a bit about those internals.

Quite frankly, I think you're actually right, we should just specify one namespace. I feel a bit dumb lol.
Wouldn't really make sense to have them in 2 different namespaces. 

I will make the change. 

PS: I noticed (but my use case might not be exhaustive) that the event might actually be used only once to transmit the subscription confirmation (not even sure) and then after that, everything goes through the subscription. Would you mind enlightening me? Do we need the event at all? Am I missing something?

Reply:
> anyone who would know a bit about those internals.

At this point, I think that's pretty much just you and me 😆 . Oh, and maybe the folks who built graphql-anycable. 

> Do we need the event at all? 

Yeah, basically the two keys are used for different kinds of communication: 

- The event key is used for taking `.trigger` calls in one Ruby process and fanning them out to other Ruby processes. - The subscription key is used for pushing specific updates from the server to the client. 

So, with the event key, the first thing that happens is a new subscription sets up a listener on that event key: 

https://github.com/rmosolgo/graphql-ruby/blob/862e8f3ee308079c4bd890b0b592bab75d224bd8/lib/graphql/subscriptions/action_cable_subscriptions.rb#L144

Then, later, when the application code calls `.trigger`, a broadcast is sent using the event key: 

https://github.com/rmosolgo/graphql-ruby/blob/862e8f3ee308079c4bd890b0b592bab75d224bd8/lib/graphql/subscriptions/action_cable_subscriptions.rb#L98-L104

The reason we use ActionCable to implement `.trigger` is because, in a multi-process Ruby deployment, there might be some subscriptions in one process, but a `.trigger` call in another process (say, a backgroud job). We need some way to notify the _other_ server process that _this_ process called `.trigger`, and so it uses ActionCable to do it. 

After that broadcast, the listeners are called, which re-evaluates the graphql query to generate an update payload:

https://github.com/rmosolgo/graphql-ruby/blob/862e8f3ee308079c4bd890b0b592bab75d224bd8/lib/graphql/subscriptions/action_cable_subscriptions.rb#L153

After that, the subscription key is used to send the update payload from the server to the client: 

https://github.com/rmosolgo/graphql-ruby/blob/862e8f3ee308079c4bd890b0b592bab75d224bd8/lib/graphql/subscriptions/action_cable_subscriptions.rb#L110

(There's no Ruby listener for that event; instead, it's a JavaScript client that we expect to be listening.)

Does that help?

Reply:
Ok, the key here is `stream_from(...) do ...`. I didn't entirely understand it as a listener and it stopped me from getting the full picture.
Now I understand, thanks for the explanation :)